Refactor TsFile

HTHou opened the pull request:
https://issues.apache.org/jira/browse/IOTDB-243

LeiRui commented:
The travis tests fail

HTHou commented:
> The travis tests fail

Yes, there are still lots of works to complete. So it's only a draft.

Ring-k commented on tsfile/src/main/java/org/apache/iotdb/tsfile/compress/ICompressor.java:
```diff
@@ -47,12 +49,12 @@ static ICompressor getCompressor(CompressionType name) {
-        return new SnappyCompressor();
-      default:
-        throw new CompressionTypeNotSupportedException(name.toString());
+    case UNCOMPRESSED:
```
Please check the indention.

Ring-k commented on tsfile/src/main/java/org/apache/iotdb/tsfile/read/controller/IMetadataQuerier.java:
```diff
@@ -49,19 +51,20 @@
+   * @throws IOException
    */
-  TSDataType getDataType(String measurement) throws NoMeasurementException;
+  TSDataType getDataType(Path path) throws NoMeasurementException, IOException;
```
The comment should be updated too.

HTHou commented:
The new structure of TsFileMetadata is shown below.
![Screen Shot 2020-02-11 at 9 31 17 AM](https://user-images.githubusercontent.com/25913899/74205308-ce29f500-4cb1-11ea-8f0b-9f5fb1859974.png)

More detailed doc to be added soon.

liutaohua commented:
I think `TsOffsetArray` is a bad design,  because it is hard to understanding it, and each `DeviceIndexArray` takes `8 byte` to store the index.
If you put the contents of `TsOffsetArray` directly into the `DeviceIndexArray`, it would only take `4 byte` to record the number, and it would be easier to understand

fanhualta commented on tsfile/src/main/java/org/apache/iotdb/tsfile/write/writer/TsFileIOWriter.java:
```diff
@@ -69,14 +67,19 @@
+  protected List<List<ChunkMetaData>> chunkGroupMetaDataList = new ArrayList<>();
+  protected List<String> deviceList = new ArrayList<>();
+  protected List<ChunkMetaData> chunkMetaDataList = new ArrayList<>();
+  private static Map<Path, List<ChunkMetaData>> timeseriesMetadataMap = new TreeMap<>();
```
Why add `static`?  I don't think all `TsFileIOWriter` instances should share this variable, which will lead to serious bugs. Please remove it.

HTHou replied:
Fixed

fanhualta commented on tsfile/src/main/java/org/apache/iotdb/tsfile/write/writer/TsFileIOWriter.java:
```diff
@@ -69,14 +67,19 @@
+  private static Map<Path, List<ChunkMetaData>> timeseriesMetadataMap = new TreeMap<>();
   private ChunkMetaData currentChunkMetaData;
   private long markedPosition;
+  private static Map<String, int[]> deviceOffsetsMap = new HashMap<>();
```
Please remove `static`. The reason is the same as above.

HTHou replied:
Fixed.

fanhualta commented on tsfile/src/main/java/org/apache/iotdb/tsfile/write/writer/TsFileIOWriter.java:
```diff
@@ -254,64 +260,51 @@ public void endFile(Schema schema) throws IOException {
+      tsOffsets[i] = out.getPosition();
+      int size = tsMetadataList.get(i).serializeTo(out.wrapAsStream());
+      tsOffsets[i+1] = tsOffsets[i] + size;
+    }
```
Please remove useless assignment and addition operations.
```
if(i > 0){
      tsOffsets[0] = out.getPosition();
}
for (i = 0; i < tsMetadataList.size(); i++) {
      int size = tsMetadataList.get(i).serializeTo(out.wrapAsStream());
      tsOffsets[i+1] = out.getPosition();
 }
```

HTHou replied:
The tsOffsets array has been removed.

fanhualta commented on tsfile/src/main/java/org/apache/iotdb/tsfile/write/writer/TsFileIOWriter.java:
```diff
@@ -254,64 +260,51 @@ public void endFile(Schema schema) throws IOException {
-   *
-   * @param chunkGroupMetaDataList all chunk group metadata in memory
-   * @return TsDeviceMetadataIndex in TsFileMetaData
+   * @return tsOffsets in TsFileMetaData
```
Please add the java doc.

HTHou replied:
Added.

fanhualta commented on tsfile/src/main/java/org/apache/iotdb/tsfile/write/writer/TsFileIOWriter.java:
```diff
@@ -254,64 +260,51 @@ public void endFile(Schema schema) throws IOException {
+      if (!deviceOffsetsMap.containsKey(deviceId)) {
+        deviceOffsetsMap.put(deviceId, new int[2]);
+        deviceOffsetsMap.get(deviceId)[0] = i;
+      }
```
Please use the following codes instead to reduce the operation and improve the efficiency.
```
deviceOffsetsMap.computeIfAbsent(deviceId, k -> new int[2])[0] = i;
```

HTHou replied:
This map is also changed.

fanhualta commented on tsfile/src/main/java/org/apache/iotdb/tsfile/write/writer/TsFileIOWriter.java:
```diff
@@ -202,13 +204,19 @@ public void writeChunk(Chunk chunk, ChunkMetaData chunkMetadata) throws IOExcept
+    List<ChunkMetaData> chunkMetaDataListOfOnePath = timeseriesMetadataMap.getOrDefault(path,
+        new ArrayList<ChunkMetaData>());
+    chunkMetaDataListOfOnePath.add(currentChunkMetaData);
+    timeseriesMetadataMap.put(path, chunkMetaDataListOfOnePath);
```
Please use the following codes instead to reduce the operation and improve the efficiency.
```
timeseriesMetadataMap.computeIfAbsent(path, k -> new ArrayList<>()).add(currentChunkMetaData);
```

HTHou replied:
Fixed.

fanhualta commented on tsfile/src/main/java/org/apache/iotdb/tsfile/write/writer/TsFileIOWriter.java:
```diff
@@ -129,54 +131,53 @@ protected void startFile() throws IOException {
    * end chunk and write some log.
    * If there is no data in the chunk group, nothing will be flushed.
    */
   public void endChunkGroup(long version) throws IOException {
```
If `version` is useless, please remove it.

HTHou replied:
Since ``version`` is used in merge part, I didn't remove it. I'll consider change this when IoTDB server works properly.

fanhualta commented on tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/ChunkGroupWriterImpl.java:
```diff
@@ -31,7 +31,7 @@
 import org.apache.iotdb.tsfile.write.record.RowBatch;
 import org.apache.iotdb.tsfile.write.record.datapoint.DataPoint;
-import org.apache.iotdb.tsfile.write.schema.MeasurementSchema;
+import org.apache.iotdb.tsfile.write.schema.TimeseriesSchema;
```
Please remove useless import.

HTHou replied:
Actually I am using the import you pointed out.

fanhualta commented on tsfile/src/main/java/org/apache/iotdb/tsfile/write/chunk/IChunkGroupWriter.java:
```diff
@@ -25,7 +25,7 @@
 import org.apache.iotdb.tsfile.write.record.RowBatch;
 import org.apache.iotdb.tsfile.write.record.datapoint.DataPoint;
-import org.apache.iotdb.tsfile.write.schema.MeasurementSchema;
+import org.apache.iotdb.tsfile.write.schema.TimeseriesSchema;
```
Please remove useless import.

fanhualta commented on tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/TimeseriesMetaData.java:
```diff
@@ -0,0 +1,117 @@
+    this.measurementId = measurementId;
+    this.chunkMetaDataList = chunkMetaDataList;
+    this.numOfChunkMetaDatas = chunkMetaDataList.size();
+  }
```
If the constructor is useless, please remove it.

HTHou replied:
Fixed. Thank you.

qiaojialin commented on tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/TimeseriesMetaData.java:
```diff
@@ -0,0 +1,117 @@
+  private int numOfChunkMetaDatas;
+
+  private String measurementId;
+  private List<ChunkMetaData> chunkMetaDataList = new ArrayList<>();
```
No need to store this in TimeseriesMetaData. I expect TimeseriesMetadata be used in this way:

List<ChunkMetadata> TsFileSequenceReader.getAllChunkMetadata(TimeseriesMetaDataIndex metadataIndex)

HTHou replied:
Thank you, fixed.

qiaojialin commented on tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/TimeseriesMetaData.java:
```diff
@@ -0,0 +1,117 @@
+
+import org.apache.iotdb.tsfile.utils.ReadWriteIOUtils;
+
+public class TimeseriesMetaData {
```
```suggestion
public class TimeseriesMetadataIndex {
```

Actually this is an index to ChunkMetadatas of one series. Similar to the TsDeviceMetadataIndex

qiaojialin commented on tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/TimeseriesMetaData.java:
```diff
@@ -0,0 +1,117 @@
+
+  private long startOffsetOfChunkMetaDataList;
+  private int chunkMetaDataListDataSize;
+  private int numOfChunkMetaDatas;
```
Do we need this field?

We could read ChunkMetadata from a ByteBuffer like this:

List<ChunkMetadata> metadatas = new ArrayList();
for (bytebuffer.remaining > 0) {
metadatas.add(bytebuffer.readChunkMetadata());
}

HTHou replied:
Fixed. Thank you.

qiaojialin commented on tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/enums/TSDataType.java:
```diff
@@ -60,20 +60,20 @@ public static int getSerializedSize() {
-        return 5;
-      default:
-        return -1;
+    case BOOLEAN:
```
Do not change this, our google-style contains the blanks.

qiaojialin commented on tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/enums/CompressionType.java:
```diff
@@ -31,22 +31,22 @@
-        return PLA;
-      default:
-        return UNCOMPRESSED;
+    case 0:
```
change back

qiaojialin commented on tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/enums/TSFreqType.java:
```diff
@@ -30,14 +30,14 @@
-        return IRREGULAR_FREQ;
-      default:
-        return IRREGULAR_FREQ;
+    case 0:
```
change back

qiaojialin commented on tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/statistics/Statistics.java:
```diff
@@ -62,33 +66,33 @@
-        return new FloatStatistics();
-      default:
-        throw new UnknownColumnTypeException(type.toString());
+    case INT32:
```
change back

qiaojialin commented on tsfile/src/main/java/org/apache/iotdb/tsfile/fileSystem/fileInputFactory/HDFSInputFactory.java:
```diff
@@ -35,19 +37,16 @@
-      logger.error(
-          "Failed to get HDFSInput in Hadoop file system. Please check your dependency of Hadoop module.",
-          e);
+      logger.error("Failed to get HDFSInput in Hadoop file system. Please check your dependency of Hadoop module.", e);
```
I think you need to use idea to format the codes......... Maybe this line exceeds the max length.

HTHou commented:
The updated structure is here. 

![Screen Shot 2020-02-12 at 4 25 20 PM](https://user-images.githubusercontent.com/25913899/74328080-3e2a9f00-4dc8-11ea-8330-bf6205fbd39f.png)

HTHou commented:
I changed some UnitTests in IoTDB server, because I added a new kind of MNode `DeviceMNode` in MTree. As a result, the deviceId cannot be storage group name any more. The storage group must be on the higher level of device.
